src/backend/joanie/core/models/products.py

```diff
@@ -389,12 +453,30 @@ def validate(self):
         Transition order to validated state.
         """
 
-    @transition(field="state", source="*", target=enums.ORDER_STATE_CANCELED)
+    @transition(
+        field="state",
+        source="*",
+        target=enums.ORDER_STATE_CANCELED,
+    )
     def cancel(self):
         """
         Mark order instance as "canceled".
         """
```

Comment:
You should remove the `order_post_transition_callback` then : 
1. In validated transition, call the `enroll_user_to_course_run` method
2. In cancel transition, call the `unroll_user_to_course_run` method